app/src/main/java/org/mozilla/focus/exceptions/ExceptionMigrationMiddleware.kt

```diff
@@ -0,0 +1,65 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.focus.exceptions
+
+import android.content.Context
+import kotlinx.coroutines.runBlocking
+import mozilla.components.browser.state.action.BrowserAction
+import mozilla.components.browser.state.action.TabListAction
+import mozilla.components.browser.state.selector.privateTabs
+import mozilla.components.browser.state.state.BrowserState
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.MiddlewareContext
+import org.mozilla.focus.components.EngineProvider
+
+/**
+ * [Middleware] implementation for migrating the tracking protection exceptions: Loading the list of
+ * exceptions from SharedPreferences and passing them to GeckoView.
+ *
+ * This is a workaround until GeckoView can permanently save tracking protection exceptions for
+ * private tabs. Otherwise they get dropped once the last GeckoSession gets closed or the process
+ * shuts down.
+ */
+class ExceptionMigrationMiddleware(
+    private val context: Context
+) : Middleware<BrowserState, BrowserAction> {
+    private val migrator by lazy { EngineProvider.provideTrackingProtectionMigrator(context) }
+    private var hasTabs = false
+
+    override fun invoke(
+        context: MiddlewareContext<BrowserState, BrowserAction>,
+        next: (BrowserAction) -> Unit,
+        action: BrowserAction
+    ) {
+        next(action)
+
+        when (action) {
+            is TabListAction -> checkIfMigrationIsNeeded(context.state)
+            else -> {}
+        }
```

Comment:
nit: Maybe `if` as we just want to identify `if(action is TabListAction)` only.

Reply:
Good point, this code went through several iterations with different numbers of actions. :)

Reply:
I just imagine how fun was that process 😄